File: behavioral/observer/observer.ts
enum States {
    eating,
    walking
}

interface Subject {
    notifyObservers(): void;
    registerObserver(observer: Observer): void;
    removeObserver(observer: Observer): void;
}

interface Observer {
    update(): void;
}

class Cat implements Subject {
    private _name: string;
    private _observers: Observer[] = [];
    private _state: States;

    constructor(name: string) {
        this._name = name;
    }

    public notifyObservers(): void {
        this._observers.forEach(observer => {
            observer.update();
        });
    }

    public registerObserver(o: Observer): void {
        this._observers.push(o);
    }

    public removeObserver(o: Observer): void {
        this._observers = this._observers.filter(observer => {
            if (observer !== o) return observer;
        });
    }

    public get name(): string {
        return this._name;
    }

    public get state(): States {
        return this._state;
    }

    public set state(state: States) {
        this._state = state;
        this.notifyObservers();
    }
}

class CatCaretaker implements Observer {
    protected _action: string;
    protected _stateToReact: States;
    protected _name: string;
    protected _observable: Cat;

    constructor(name: string, action: string, stateToReact: States) {
        this._name = name;
        this._action = action;
        this._stateToReact = stateToReact;
    }

    public registerAsObserver(observable: Cat): void {
        if (this._observable) return;

        observable.registerObserver(this);
        this._observable = observable;
    }

    public removeAsObserver(): void {
        if (!this._observable) return;

        this._observable.removeObserver(this);
        this._observable = null;
    }

    public update(): void {
        if (this._observable && this._observable.state === this._stateToReact) this._react();
    }

    protected _react(): void {
        console.log(`${this._observable.name} is being ${this._action} by ${this._name}.`);
    }
}

class Dog implements Observer {
    public update(): void {
        console.log('The dog has looked at the cat.');
    }
}

let cat: Cat = new Cat('Simba');
let feeder: CatCaretaker = new CatCaretaker('Feeder', 'fed', States.eating);
let walker: CatCaretaker = new CatCaretaker('Walker', 'walked', States.walking);
let dog: Dog = new Dog();

feeder.registerAsObserver(cat);
walker.registerAsObserver(cat);
cat.registerObserver(dog);

cat.state = States.eating;
cat.state = States.walking;
cat.removeObserver(dog);
cat.state = States.eating;

File: behavioral/observer/observer.cs
using System;
using System.Collections.Generic;
using System.Linq;

class ObserverPattern
{
    enum States
    {
        eating,
        walking
    }

    interface Subject
    {
        void NotifyObservers();
        void RegisterObserver(Observer observer);
        void RemoveObserver(Observer observer);
    }

    interface Observer
    {
        void Update();
    }

    class Cat : Subject
    {
        private States state;
        private List<Observer> observers = new List<Observer>();

        public string name { get; }
        public States State
        {
            get
            {
                return state;
            }
            set
            {
                state = value;
                NotifyObservers();
            }
        }

        public Cat(string name)
        {
            this.name = name;
        }

        public void NotifyObservers()
        {
            observers.ForEach(observer => {
                observer.Update();
            });
        }

        public void RegisterObserver(Observer o)
        {
            observers.Add(o);
        }

        public void RemoveObserver(Observer o)
        {
            observers = observers
                .Where(observer => observer != o)
                .ToList();
        }
    }

    class CatCaretaker : Observer
    {
        protected string name;
        protected Cat observable;
        protected string action { get; }
        protected States stateToReact { get; }

        public CatCaretaker(string name, string action, States stateToReact)
        {
            this.name = name;
            this.action = action;
            this.stateToReact = stateToReact;
        }

        public void RegisterAsObserver(Cat observable)
        {
            if (this.observable != null) return;

            observable.RegisterObserver(this);
            this.observable = observable;
        }

        public void RemoveAsObserver()
        {
            if (observable == null) return;

            observable.RemoveObserver(this);
            observable = null;
        }

        public void Update()
        {
            if ((observable?.State == stateToReact)) React();
        }

        protected void React()
        {
            Console.WriteLine("{0} is being {1} by {2}.", observable.name, action, name);
        }
    }

    class Dog : Observer
    {
        public void Update()
        {
            Console.WriteLine("The dog has looked at cat.");
        }
    }

    static void Main(string[] args)
    {
        Cat cat = new Cat("Simba");
        CatCaretaker feeder = new CatCaretaker("Feeder", "fed", States.eating);
        CatCaretaker walker = new CatCaretaker("Walker", "walked", States.walking);
        Dog dog = new Dog();

        feeder.RegisterAsObserver(cat);
        walker.RegisterAsObserver(cat);
        cat.RegisterObserver(dog);

        cat.State = States.eating;
        cat.State = States.walking;
        cat.RemoveObserver(dog);
        cat.State = States.eating;
    }
}
